addons/account/models/account_account.py

```diff
@@ -516,8 +521,13 @@ def _search_new_account_code(self, start_code, cache=None):
         if cache is None:
             cache = {start_code}
 
+        company_ids_to_check = (self.env.company.child_ids | self.env.company.parent_ids).ids
+
         def code_is_available(new_code):
-            return new_code not in cache and not self.search_count([('code', '=', new_code)], limit=1)
+            return (
+                new_code not in cache
```

Comment:
The cache needs to be per company too then, no? Might be too complicated for a corner case though

Reply:
Codes are per-root-company, and we are only checking accounts that belong to child companies of the same root company, so we don't need several caches.

Reply:
(I think we discussed this in person, but writing this just so it's written: in the create, we reinitialize `cache` for every `company_ids`, and we do `env['account.move'].with_company(companies[0])._search_new_account_code()`, so effectively the cache is per company)